### ISSUE TITLE
[REVIEW]: SenSARP: A Python package for pre-processing Sentinel-1 SLC data with the Sentinel-1 Toolbox

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3337]
**Submitting author:** @McWhity (<a href="http://orcid.org/0000-0001-5278-8379">Thomas Weiß</a>)
**Repository:** <a href="https://github.com/multiply-org/sar-pre-processing.git" target ="_blank">https://github.com/multiply-org/sar-pre-processing.git</a>
**Version:** v0.2
**Editor:** @pdebuyl
**Reviewers:** @arthur-e, @danclewley
**Archive:** <a href="https://doi.org/10.5281/zenodo.5903483" target="_blank">10.5281/zenodo.5903483</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/4d1763625f0fe33552e277d89ac18049/status.svg)](https://joss.theoj.org/papers/4d1763625f0fe33552e277d89ac18049)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/4d1763625f0fe33552e277d89ac18049"><img src="https://joss.theoj.org/papers/4d1763625f0fe33552e277d89ac18049/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/4d1763625f0fe33552e277d89ac18049/status.svg)](https://joss.theoj.org/papers/4d1763625f0fe33552e277d89ac18049)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/multiply-org/sar-pre-processing.git" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@arthur-e & @braunfuss, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @pdebuyl know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/multiply-org/sar-pre-processing.git">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@McWhity) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/multiply-org/sar-pre-processing.git">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@McWhity) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @arthur-e, @braunfuss it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
Failed to discover a `Statement of need` section in paper

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.09 s (531.8 files/s, 54898.0 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
XML                              4              7              0            983
reStructuredText                11            253             80            820
Python                          10            257            347            813
YAML                            10             39             47            314
make                             1             28              6            143
Markdown                         3             40              0            107
Jupyter Notebook                 1              0            274             75
TeX                              1              8              0             73
Bourne Shell                     2              1              0             24
HTML                             1              1              0              6
DOS Batch                        1              0              0              2
JSON                             1              0              0              1
-------------------------------------------------------------------------------
SUM:                            46            634            754           3361
-------------------------------------------------------------------------------


Statistical information for the repository '0325dfefdfd695f9447bbdb2' was
gathered on 2021/06/07.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
McWhity                         47          3004           1755           67.12
Thomas Weiß                     62           457            432           12.54
Tonio Fincke                    39           764            634           19.72
TonioF                           9            27             17            0.62

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Thomas Weiß                1114          243.8         30.1               23.52
Tonio Fincke                303           39.7         16.1                4.62
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03337/joss.03337/10.21105.joss.03337.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03337/joss.03337/10.21105.joss.03337.pdf) :page_facing_up: :point_left:

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.3390/rs10081286 is OK
- 10.3390/w11091938 is OK
- 10.3390/rs12183037 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by pdebuyl** [comment]
@arthur-e @braunfuss make sure to accept the invitation to the reviewers group and to have a look at the reviewer guidelines linked to at the top of this review page.

@McWhity JOSS requires that a section of the article is literally titled "statement of need", can you edit the article accordingly?

The review process will happen in this issue page, so questions to the author or to me can be added as comments here. As this is the first JOSS review for both reviewers, do not hesitate to ask questions if you have doubts about the procedure. You can also take a look at earlier reviews to get an idea of how the reviews proceed https://github.com/openjournals/joss-reviews/issues?q=is%3Aissue+label%3Aaccepted+

**Comment by pdebuyl** [comment]
@arthur-e @braunfuss do you need any assistance for starting the review ?

**Comment by arthur-e** [comment]
@pdebuyl I think I have everything I need, thanks!

**Comment by whedon** [comment]
:wave: @braunfuss, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @arthur-e, please update us on how your review is going (this is an automated reminder).

**Comment by pdebuyl** [comment]
Hi @arthur-e thank you very much for the review. With JOSS, we do not expect a "minor/major revision" assessment. Providing a detailed report, as you have done, participates in the constructive review process which will take place in this issue.

**Comment by arthur-e** [comment]
@pdebuyl Thanks! [The "Review criteria"](https://joss.readthedocs.io/en/latest/review_criteria.html#what-happens-if-the-software-i-m-reviewing-doesn-t-meet-the-joss-criteria) may need to be updated as it still mentions "Minor" and "Major" revisions.

**Comment by pdebuyl** [comment]
Thank you @arthur-e for pointing this out. In practice, the reviews proceed by iterative improvements and I forgot that we request this grading. I'll discuss it with the editor group.

**Comment by pdebuyl** [comment]
@McWhity you can already address @arthur-e 's review (without waiting for @braunfuss ' one) or wait for the second review to arrive.

**Comment by McWhity** [comment]
> @McWhity you can already address @arthur-e 's review (without waiting for @braunfuss ' one) or wait for the second review to arrive.

Yes I will address the review in the next view days

**Comment by pdebuyl** [comment]
Hi @braunfuss any timeline for the review?

**Comment by McWhity** [comment]
Unfortunately, I haven't found time to address the detailed review of @arthur-e yet. I think I will be able to update the repository based on the review of @arthur-e until 22nd of August. This means in terms of saving work time for the reviewers, @braunfuss might wait with conducting his review until the 22nd of August. If @braunfuss is almost finished with his review, I will be happy to include his remarks too.

**Comment by pdebuyl** [comment]
Hi @McWhity is the update coming as planned?

**Comment by McWhity** [comment]
Sorry, I'm a little bit behind schedule. I'm currently testing the implemented changes. Therefore, the new version should be ready within this week. I will keep you posted when I push the final changes. Furthermore, I will provide a detailed answer to the review comments.

**Comment by McWhity** [comment]
> I have completed my review.
> 
> I am unsure whether the submission requires "Minor" or "Major" revisions. I would tentatively propose that "Major revisions" are required only because I could not fully evaluate the functionality of the software due to an outstanding technical issue.
> ## Comments to Authors
> 
> Overall, SenSARP seems to provide a valuable, very high-level processing chain for Sentinel 1 SLC data. However, there are two main areas in need of improvement: 1) Key functionality issue(s) need to be resolved; and, 2) The documentation should be improved and expanded to provide more details to users that may want to deviate from the prescribed pre-processing chain.
> 
> **First, there is [an unresolved issue](https://github.com/multiply-org/sar-pre-processing/issues/12) as to the functionality of pre-processing Step 1.** While unresolved, this issue prevents the software from functioning and, therefore, impedes further evaluation of its functionality.
> 
> I'd like to see the Documentation updated to provide end-users with more information on how to customize their processing chain. In addition to adding more details on the end-user functionality, there are a number syntax/ typographical issues with the documentation (e.g., missing citations in the ReadTheDocs compilation) that should fixed. Also, while SenSARP seems to have automated tests (and CI through Travis CI), the authors should add instructions to the README on how users can run tests themselves. I was able to run the two Python scripts in the `tests/` directory without error, but I didn't get any feedback in the console so I can't be sure. I've written more about proposed Documentation improvements in the section on Documentation, below.
> 
Thank you for pointing this out.
- Syntax/typographical issues (missing citations, ReadTheDocs compilation) within documentation are resolved
- Documentation is updated to provide more end-user information (see answers to review points on the documentation)
- Information how to run test within [Readme](https://github.com/multiply-org/sar-pre-processing/blob/master/README.md) and [ReadtheDocs documentation](https://multiply-sar-pre-processing.readthedocs.io/en/latest/01_Introduction.html) were added


> It is also difficult to determine how much the SenSARP Toolbox improves upon the ESA SNAP Sentinel-1 Toolbox. For instance, [`SARPreProcessor.pre_process_step1()`](https://multiply-sar-pre-processing.readthedocs.io/en/latest/_modules/sar_pre_processing/sar_pre_processor.html#SARPreProcessor.pre_process_step1) consists of reading a configuration file, setting defaults, and using `subprocess.call()` to activate the corresponding SNAP Sentinel-1 Toolbox functions. The authors should revise the accompanying paper as well as the documentation to more explicitly state how this software improves upon the SNAP toolbox. If the major improvement is that `SenSARP` offers a programmatic API to the SNAP toolbox functions, that would be a meaningful contribution, but it should be made clear. While the authors do make the case that "an automatic or manual pre-processing of Sentinel-1 images is needed" (Lines 19-20), it would be good to know for sure that the SNAP toolbox _does not_ offer such an automated pipeline, so that the value of `SenSARP` is obvious. **In short, what are the relative advantages of SenSARP over the SNAP Toolbox, which is a dependency?** I believe answering this question will better address the journal requirement: "A Statement of Need section that clearly illustrates the research purpose of the software." The journal recommends a section titled "Statement of Need" in the paper, though I think a section header is not necessary as long as the need is clearly described towards the beginning of the paper.
> 
> I think that SenSARP could be characterized differently in the JOSS paper and in the software's documentation/ README to make it clear that it provides a specific pre-processing chain with sensible(?) defaults--I would not call it a "toolbox" myself, [because it has very few individual functions/ classes for end-users](https://multiply-sar-pre-processing.readthedocs.io/en/latest/annex_technical.html), but I do not dispute the use of that term. I think it might be more useful (and accurate?) to characterize `SenSARP` as a "push-button" or "turn-key" solution (or some other analogy) to automated pre-processing of Sentinel-1 SLC data. It seems that rather than a multi-functional programming API, `SenSARP` makes it easy to complete a complex but fairly rigid processing chain to multiple images. My first impression of `SenSARP` was that it would have multiple end-points/ functions/ classes for representing different kinds of data and processing steps; but, [it seems to consist primarily of three end-user functions](https://multiply-sar-pre-processing.readthedocs.io/en/latest/annex_technical.html), `pre_process_step1()`, `pre_process_step2()`, `pre_process_step3()`. So, I think of `SenSARP` more as a pipeline rather than a "toolbox."
> 
Thank you for these very helpful remarks. A section "Statement of Need" addressing the mentioned shortcomings is now added to the [Paper](https://github.com/multiply-org/sar-pre-processing/blob/master/docs/paper.pdf), the [Readme](https://github.com/multiply-org/sar-pre-processing/blob/master/README.md), and the [Documentation](https://multiply-sar-pre-processing.readthedocs.io/en/latest/01_Introduction.html). SenSARP is not characterized as "toolbox" anymore but as suggested SenSARP is now called a "push-button" / "pipeline" solution

> Finally, the authors should address the journal requirement on Community Guidelines: "There should be clear guidelines for third-parties wishing to: Contribute to the software; Report issues or problems with the software; Seek support." Some repositories include a `CONTRIBUTING` (Markdown or RST) file with detailed instructions about how Pull Requests and Issues should be formatted. Alternatively, this information could be added to a section of the `README`.

Thank you for pointing this out. A section "Support, contributing and testing" was added to the [Readme](https://github.com/multiply-org/sar-pre-processing/blob/master/README.md) and the [documentation](https://multiply-sar-pre-processing.readthedocs.io/en/latest/01_Introduction.html)

> ### Functionality
> 
> **As mentioned above, there is [an unresolved issue](https://github.com/multiply-org/sar-pre-processing/issues/12) as to the functionality of pre-processing Step 1.**
> 
Please see answer to [an unresolved issue](https://github.com/multiply-org/sar-pre-processing/issues/12)

> [Another outstanding issue](https://github.com/multiply-org/sar-pre-processing/issues/13) that is unrelated may be a problem with the ESA data service and not with SenSARP. However, this issue (with the ESA data service) seems to severely impact part of the SenSARP functionality, in that data must be downloaded manually from somewhere else. The authors should address this shortcoming and provide some guidance to end-users in the Documentation and the Jupyter Notebook. **Is there somewhere else users can get Sentinel 1 SLC data?** I used NASA Earth Data Search.
> 
Thank you for this remark. You are right SenSARP can only be used if the Sentinel-1 data is available. Two options are now introduced within the [jupyter notebook](https://github.com/multiply-org/sar-pre-processing/blob/master/docs/notebooks/default_process_single_image.ipynb).
- Option 1: Downloading data manually or via sentinelsat from Copernicus data hub.
- Option 2: Downloading data from the Alaska Satellite Facility (ASF uses the same credentials als NASA Earth Data Search).
The two options are also introduced within the [documentation](https://multiply-sar-pre-processing.readthedocs.io/en/latest/notebooks/default_process_single_image.html)

> **Regarding Installation...**
> 
>     * **One key change that is needed: Version details for the SNAP Toolbox dependency.** In the review, [it became apparent that Version 8.0.4 (or greater?) is required](https://github.com/multiply-org/sar-pre-processing/issues/12#issuecomment-863323367) but in the installation instructions there is no such information--and an unsupported version will be installed by default when the SNAP Toolbox is downloaded. Users should be told explicitly (how) to upgrade to 8.0.4.
> 
Thank you for this comment. Information that a update is needed as well as information how to update SNAP is given within the [Readme](https://github.com/multiply-org/sar-pre-processing/blob/master/README.md), the beginning of the [jupyter notebook](https://github.com/multiply-org/sar-pre-processing/blob/master/docs/notebooks/default_process_single_image.ipynb) and the installation part of the [documentation](https://multiply-sar-pre-processing.readthedocs.io/en/latest/02_Installation.html#further-information)

>     * I would prefer it if there were some way to install the software without using `conda`, which can be a pain to work with (and very slow). In this case, `conda` took several minutes to install the repository (spent a lot of time on "Solving environment"). Ideally, the `setup.py` script would list the package dependencies so that the repository (and its dependencies) could be installed with either `distutils` or `pip`. This can still be done within a `conda` virtual environment or, better yet, a `virtualenv` environment.
> 
>     * It would be helpful if the installation instructions could describe/ link to `conda` installation instructions, both for Anaconda and `miniconda`.
> 
>     * A few more details on how to download and install the ESA SNAP Sentinel toolbox would be helpful to users.

Thank you for this remark.
- Instruction how to install SenSARP via virtualenv and python were added. As the installation of GDAL can be very tricky, GDAL needs to be installed separately. Instruction how to do this is were added
- Information how to download and install Anaconda and Miniconda were added
- More details on how to download and install the ESA SNAP Sentinel toolbox were added

Please see installation part of [readme](https://github.com/multiply-org/sar-pre-processing/blob/master/README.md) and [documentation](https://multiply-sar-pre-processing.readthedocs.io/en/latest/02_Installation.html#installation-with-virtualenv-and-python)

> 
> 
> ### Documentation
> 
> [In the "Usage" section,](https://multiply-sar-pre-processing.readthedocs.io/en/latest/notebooks/running_test_application.html#2.-Download-sample-data-from-Sentinel-Data-Hub) there seem to be some errors that occurred at the time the documentation was compiled (e.g., `ModuleNotFoundError: No module named 'sentinelsat'`). Fixing these and showing the correct output should help users to understand what the correct behavior of the software looks like.
> 
Thank you for pointing this out. The ["Usage"](https://multiply-sar-pre-processing.readthedocs.io/en/latest/03_Usage.html) section was revised. Now showing three jupyter notebook examples (default processing of a single image, default processing a time series, processing a time series with a user defined xml graph) and an explanation of the config file.

> I really like the detailed description of each processing step, however, the "Practical implementation" sections seem incomplete. While most users might use the defaults for `pre_process_step1()`, `pre_process_step2()`, and `pre_process_step3()`, is it possible that a user might want to skip or change a step? For instance, maybe a user wants to skip radiometric correct in this pipeline and do it differently later. Backscatter normalization is also "optional" [but it's not clear from the documentation](https://multiply-sar-pre-processing.readthedocs.io/en/latest/04_Theory.hhttps://multiply-sar-pre-processing.readthedocs.io/en/latest/04_Theory.html#backscatter-normalisation-optionaltml#backscatter-normalisation-optional) how to turn it on/ off. I would like to see these details fully described in the "Practical implementation" sections, perhaps with a minimal example (of how to configure the YAML file for that aim?).
> 
Thank you for pointing this out. As the theory of default pre-processing steps are explained within "Practical implementation" the section was renamed to "Default processing chain". For non-expert users skipping of a step is not intended right now. I understand the confusion about the "optional" behind the "backscatter normalization" step. The backscatter normalization as well as the multi-temporal filter are applied in any case. However, the final netcdf file contains layers of
- single speckle filtered radiometric and geometric corrected sigma nought backscatter
- multi speckle filtered radiometric and geometric corrected sigma nought backscatter
- single speckle filtered radiometric and geometric corrected sigma nought backscatter normalized to a specific incidence angle
- multi speckle filtered radiometric and geometric corrected sigma nought backscatter normalized to a specific incidence angle

The missing information about which layers and which content is provided by the final netcdf file a section explaining the [output layers](https://multiply-sar-pre-processing.readthedocs.io/en/master/04_Theory.html#output-layers-of-final-netcdf-stack-file) was added. Another section ["Folder and data creation during pre-processing steps"](https://multiply-sar-pre-processing.readthedocs.io/en/master/04_Theory.html#folder-and-data-creation-during-pre-processing-steps) explaining which folders are created during the default pre-processing was added as well.
Regarding the skipping of specific processing steps. A new [jupyter notebook](https://github.com/multiply-org/sar-pre-processing/blob/master/docs/notebooks/use_user_defined_graphs.ipynb) explaining how expert users can modify the default xml-processing graphs or create their own processing graphs and use the SenSARP functionalities was added to the ["Usage"](https://multiply-sar-pre-processing.readthedocs.io/en/master/notebooks/use_user_defined_graphs.html#3.-Use-user-defined-xml-graph-to-process-Sentinel-1-data-(test-case-with-Sentinel-1-GRD-product)) section.

> There are also some minor syntax issues with the documentation that could be fixed, e.g., [the Theory/purpose section of the section on TOPSAR Deburst](https://multiply-sar-pre-processing.readthedocs.io/en/latest/04_Theory.html#id12) reads: "More detailed information are provided in [], [] and []." I think that _none_ of the in-line citations are appearing on ReadTheDocs for any section.
> 
Thank you for pointing this out. Syntax issues were fixed.


> Finally, I think the README could be updated to provide explicit instructions to users for running the tests, including information as to how to evaluate they ran correctly--for instance, the users should see _no output_ in the console if the tests succeeded, correct?

Thank you for pointing this out. As mentioned previously a section "Support, contributing and testing" was added. If you run "pytest" in the terminal you should receive an output if the test did run correctly or if they failed.

> ### Software Paper
> 
> As discussed above, I believe the authors could do more to distinguish SenSARP from the SNAP Toolbox. Specifically:
> 
>     * How does SenSARP improve uponthe SNAP Toolbox? It would appear that SenSARP provides [the pre-processing functionality of the SNAP Toolbox](http://step.esa.int/main/toolboxes/sentinel-1-toolbox/s1tbx-features/) via `subprocess` calls to SNAP. This may be a substantial improvement over the graphical user interface (GUI) of SNAP through the ability to automate these steps at scale. If this is the primary contribution of SenSARP, the authors should explicitly state so in the paper.
> 
Thank you for your very helpful remarks. As already mentioned a section "Statement of need" clarifying the mentioned shortcomings was added. Another feedback in this matter is highly appreciated.

>     * Is SenSARP a fairly rigid pre-processing pipeline or is it a multi-functionality API for creating customized GPT pipelines? Neither the documentation nor the Jupyter Notebook provide examples of how the pipeline might be customized to skip or change certain pre-processing steps.

Thank you for your very helpful remarks. A new [juypter notebook](https://multiply-sar-pre-processing.readthedocs.io/en/latest/notebooks/use_user_defined_graphs.html#3.-Use-user-defined-xml-graph-to-process-Sentinel-1-data-(test-case-with-Sentinel-1-GRD-product)) explaining how expert users can customize or use their own pre-processing chain was added.

**Comment by pdebuyl** [comment]
Hi @McWhity , sorry I had missed your replies. Thank you for addressing the comments by @arthur-e .

**Comment by pdebuyl** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03337/joss.03337/10.21105.joss.03337.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03337/joss.03337/10.21105.joss.03337.pdf) :page_facing_up: :point_left:

**Comment by pdebuyl** [comment]
Hi @McWhity for information, I have started looking for an alternative second reviewer.

**Comment by McWhity** [comment]
Thank you for the update

**Comment by pdebuyl** [comment]
@whedon add @danclewley as reviewer

**Comment by whedon** [comment]
OK, @danclewley is now a reviewer

**Comment by pdebuyl** [comment]
@whedon remove @braunfuss as reviewer

**Comment by whedon** [comment]
OK, @braunfuss is no longer a reviewer

**Comment by pdebuyl** [comment]
Hi @danclewley thank you for accepting to review. The general guidelines for reviewing are here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html

In a nutshell:
- Read our [review criteria](https://joss.readthedocs.io/en/latest/review_criteria.html).
- Take a look at the package, install it, check and run the tests and examples.
- Go over your checklist (in the first message of this page) and use it as a guide to perform the review.

Don't forget to accept the invitation to the group of reviewers, you need to accept it to be able to check the items in the checklist.

**Comment by pdebuyl** [comment]
Hi @danclewley thank you for getting on to this.

Regarding scholarly efforts, you can find the criteria here: https://joss.readthedocs.io/en/latest/review_criteria.html#substantial-scholarly-effort

During the pre-review phase, during which the question was raised by Daniel Katz, the author already provided extra information https://github.com/openjournals/joss-reviews/issues/3198#issuecomment-841809933

Among the criteria for scholary effort, the paper mentions existing applications of the package, which also count toward the evalution. I let the author reply for further motivation if necessary.

**Comment by danclewley** [comment]
Apologies this has taken so long.

I'm happy to recommend **accept with minor revisions** if the paper is updated to clarify who it is aimed at and how it differs from other packages.

**Comment by pdebuyl** [comment]
Hi @danclewley thank for the feedback.

To proceed, I would like to ask you to fill the checklist for all items that you agree on. This is part of the JOSS review process and it in general also a nice way for the authors to track the progress of the review. To accept the paper, once you are ok, I would also like an explicit statement of acceptance here. (This is not part of JOSS guidelines but without this, with the back and forth messages and the absence of notifications for checklist editing, I find it harder to conclude the review).

**Comment by pdebuyl** [comment]
@arthur-e would you update your checklist as well in light of the updates?

**Comment by arthur-e** [comment]
@pdebuyl I am satisfied that the authors have revised the manuscript, the software, and the documentation to meet the journal's standards. I have updated my checklist to reflect that.

**Comment by pdebuyl** [comment]
Thanks a lot for finishing the review @arthur-e

**Comment by pdebuyl** [comment]
Hi @McWhity can you reply to @danclewley 's review?

**Comment by McWhity** [comment]
Sorry for the delay. I will give an update/response on the weekend.

@arthur-e and @danclewley: Thanks a lot for the review.

**Comment by McWhity** [comment]
> Apologies this has taken so long.
> 
> I'm happy to recommend **accept with minor revisions** if the paper is updated to clarify who it is aimed at and how it differs from other packages.

@danclewley: Thank you for the review. I updated the paper (https://github.com/multiply-org/sar-pre-processing/blob/master/docs/paper.pdf).

Main changes are:
- last paragraph of method section was rewritten:
> After the pre-processing the resulting radiometrically and geometrically corrected images are stored for further usage within a NetCDF4 stack file. The processing workflow was developed and optimized to use a Sentinel-1 time series of pre-processed sigma nought backscatter values to retrieve biophysical land surface parameters by the use of radiative transfer models. The sigma nought backscatter values provided by the default workflow of SenSARP might be used in other applications like flood risk analysis, monitoring land cover changes or monitoring global food security but it has to be mentioned that different applications have different demands and therefore, slight adjustments of the default workflow might be required. In the future, many more new products and operational third party services based on consistent Sentinel-1 time series might be developed._

- new Section regarding other python packages was added:

> **Other available python software packages using ESA's SNAP software to pre-process SAR data**
The ESA's SNAP toolbox has been written in Java. For python users the developers provide a python interface called Snappy. However, the Snappy interface is lacking in terms of installation, processing performance and usability. Hence, the remote sensing community developed different wrappers (e.g. SenSARP, snapista or pyroSAR) to use SNAP processing functionalities by utilizing the SNAP Graph Processing Tool (GPT).

> **snapista**
Snapista (https://snap-contrib.github.io/snapista/index.html) targets mainly experts remote sensing users with python programming skills. It provides access to the processing operators of all toolboxes (e.g. Sentinel-1, Sentinel-2 or Sentinel-3) within SNAP. Expert users can generate processing graphs and execute there generated graphs in a pure Pythonic way. A guideline which processing steps are needed for different applications or which processing steps can or have to be combined are not provided yet. As guidelines how to process different satellite data for different applications is not an easy task to do it exceeds the goal of snapista as a python wrapper for the SNAP software. Summarizing, snapista provides access to all SNAP toolboxes (not just to Sentinel-1 Toolbox) via python. But as it provides no default processing chains, snapista will be primarily usable by expert remote sensing users. The advantage of snapista is the accessibility of processing operators for SAR and optical data.

> **pyroSAR**
PyroSAR (https://pyrosar.readthedocs.io/en/latest/index.html) is a python library which provides a python wrapper to SAR pre-processing software SNAP and GAMMA [@wegnuller_sentinel-1_2016; @werner_gamma_2000]. The library provides utilities to read and store metadata information of downloaded satellite data within a database. Furthermore, pyroSAR provides access to processing operators of SNAP and GAMMA. A default workflow with different user options is provided to process single or time-series Sentinel-1 images. After executing the default processing workflow radiometric and geometric corrected gamma nought backscatter data are provided in Geotiff format [@truckenbrodt_pyrosar_2019]. The processed images can also be stored within an Open Data Cube. For expert users which might want to use a different processing workflow pyroSAR provides an option to create SNAP xml-workflows and execute them with the GPT. Summarizing, pyroSAR provides a similar push-button option to process Sentinel-1 data with a slightly different default workflow (pyroSAR: no temporal speckle filter, gamma nought backscatter output in Geotiff format) than SenSARP (SenSARP: temporal speckle filter, sigma nought backscatter output in netCDF format). PyroSAR, as a more complex library than SenSARP, provides on the one hand more changeable parameters within the processing workflow but on the other hand the usability for non-expert users might be narrowed compared to SenSARP. An advantage of SenSARP, especially for non-expert users, might be the provision of background information (theory/purpose) of the different pre-processing steps within the documentation.

I hope these changes satisfying your request

**Comment by pdebuyl** [comment]
Thank you @McWhity for the update.

**Comment by pdebuyl** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03337/joss.03337/10.21105.joss.03337.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03337/joss.03337/10.21105.joss.03337.pdf) :page_facing_up: :point_left:

**Comment by pdebuyl** [comment]
@danclewley can you take @McWhity 's update in your review? Please make sure to tick the appropriate checkboxes at the top of this page as appropriate.

**Comment by danclewley** [comment]
Updates look good to me and address my comments. Happy to recommend accepting.

**Comment by pdebuyl** [comment]
Hi @danclewley thanks for the feedback! Can you check all relevant checkboxes? This is part of the JOSS process and despite your message I don't think it would be right for me to check them :-)

**Comment by pdebuyl** [comment]
Thanks @danclewley for filling in the form.

**Comment by pdebuyl** [comment]
Hi @McWhity the review is almost done, but as there was some issue for installing it for @danclewley I am checking with the board for a way out, by inviting a third reviewer for these steps for instance. I'll keep you updated.

**Comment by pdebuyl** [comment]
Hi @McWhity I will perform the remaining checks myself so that we can proceed.

**Comment by pdebuyl** [comment]
I have a minor comment for the installation: when using pip, it seems that you need to pick a pygdal version that is of same "point release" (in `x.y.z` only `x` and `y` need to match) so that the recommended methods can lead to incomprehension from users. Of course, nowadays using pip for scientific packages is more of an expert choice unfortunately.

**Comment by McWhity** [comment]
The installation via virtualenv and pip was added due to a comment by arthur-e.

> - I would prefer it if there were some way to install the software without using conda, which can be a pain to work with (and very slow). In this case, conda took several minutes to install the repository (spent a lot of time on "Solving environment"). Ideally, the setup.py script would list the package dependencies so that the repository (and its dependencies) could be installed with either distutils or pip. This can still be done within a conda virtual environment or, better yet, a virtualenv environment.

The installation of a working gdal version can be tricky. If using pip, `x.y.z` of the pygdal version (`x.y.z.buildnumber)` need to be the same as the previously installed `gdal-bin x.y.z` package. Does this clarify your concerns?

Beside the installation via pip and virtualenv a second installation guide by using Anaconda/Miniconda and conda is given in the documentary.

**Comment by arthur-e** [comment]
@McWhity Yes, that's great. Giving users options is always helpful. Power users will be able to install the GDAL Python bindings ahead of time and then use whatever package manager they want for the rest. Less experienced users will have to wait on Conda.

**Comment by pdebuyl** [comment]
Hi @McWhity 

- I filed a pull request for minor edits to the article. I think that the changes do not modify the intent of your text, let me know if this is the case.
- Can you archive the software on zenodo and provide here the version information of that archive and zenodo's DOI ?
- The last reference has no URL or DOI, could provide a stronger identification?

Once those items are checked, I'll forward the article for acceptance.

Regarding the installation: I admit that the specific annoyance of GDAL install goes beyond the scope of SenSARP, thanks for the clarification. I could install with the debian-provided GDAL in the end, which is my default choice :-)

**Comment by McWhity** [comment]
Hi @pdebuyl
Thank you for the remaining checks. I'm happy to check your pull request, but unfortunately don't see the pull request.

**Comment by pdebuyl** [comment]
I must have missed a clic. Fortunately, the patch was still saved by GitHub's web interface, I opened the PR.

**Comment by McWhity** [comment]
- PR is merged. Thank you for the revision
- last reference was revised (URL was added)
- changelog was updated and new tag v0.2 was created

As the Github repository is within a Github organization I need permission from the organization owner to grant zenodo access. As soon as the permission is granted, I will archive SenSARP on zenodo and provide the required information. This shouldn't take long.

**Comment by pdebuyl** [comment]
Hi @McWhity thanks for the update. I am waiting for the zenodo info then.

**Comment by McWhity** [comment]
@pdebuyl: SenSARP is now archived on zenodo. The DOI is [10.5281/zenodo.5903483](https://zenodo.org/record/5903484) (Version v0.2).

**Comment by pdebuyl** [comment]
@whedon set v0.2 as version

**Comment by whedon** [comment]
OK. v0.2 is the version.

**Comment by pdebuyl** [comment]
@whedon set 10.5281/zenodo.5903483 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.5903483" target="_blank">10.5281/zenodo.5903483</a> is the archive.

**Comment by pdebuyl** [comment]
Thank you @McWhity for the update

**Comment by pdebuyl** [comment]
@whedon recommend-accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.3390/rs10081286 is OK
- 10.3390/w11091938 is OK
- 10.3390/rs12183037 is OK
- 10.1016/j.procs.2016.09.246 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/2908

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/2908, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by pdebuyl** [comment]
Note to the EiC: I accepted to perform the functionality check in place of the second reviewer for availability reasons.

**Comment by arfon** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1487216102446878720) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/2912
1. Wait a couple of minutes, then verify that the paper DOI resolves [https://doi.org/10.21105/joss.03337](https://doi.org/10.21105/joss.03337)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by arfon** [comment]
[@arthur-e](https://github.com/arthur-e), [@danclewley](https://github.com/danclewley) – many thanks for your reviews here and to @pdebuyl for editing this submission! JOSS relies upon the volunteer effort of people like you and we simply wouldn't be able to do this without you ✨ 

@McWhity – your paper is now accepted and published in JOSS :zap::rocket::boom:

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.03337/status.svg)](https://doi.org/10.21105/joss.03337)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.03337">
  <img src="https://joss.theoj.org/papers/10.21105/joss.03337/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.03337/status.svg
   :target: https://doi.org/10.21105/joss.03337
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.03337/status.svg)](https://doi.org/10.21105/joss.03337)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss

**Comment by pdebuyl** [comment]
Thanks @arfon for finalizing this!

**Comment by McWhity** [comment]
@arthur-e and @danclewley: Thank you very much for your reviews. It helped a lot to improve the python package.
@pdebuyl and @arfon: Thanks for editing and finalizing the paper. Keep up with good work!
